src/hotspot/cpu/ppc/gc/shared/barrierSetNMethod_ppc.cpp

```diff
@@ -23,18 +23,110 @@
+    // Patching is not atomic.
+    // Stale observations of the "armed" state is okay as invoking the barrier stub in that case has no
+    // unwanted side effects. Disarming is thus a non-critical operation.
+    // The visibility of the "armed" state must be ensured by safepoint/handshake.
+
+    OrderAccess::release(); // Release modified oops
+
+    // Set the guard value (naming of 'offset' function is misleading).
+    get_patchable_instruction_handle()->set_offset(value);
+  }
+
+  void verify() const {
+    // Although it's possible to just validate the to-be-patched instruction,
+    // all instructions are validated to ensure that the barrier is hit properly - especially since
+    // the pattern used in load_const32 is a quite common one.
+
+    uint* current_instruction = reinterpret_cast<uint*>(get_barrier_start_address());
+
+    // calculate_address_from_global_toc (compound instruction)
+    verify_op_code_manually(current_instruction, MacroAssembler::is_addis(*current_instruction));
+    verify_op_code_manually(current_instruction, MacroAssembler::is_addi(*current_instruction));
+
+    verify_op_code_manually(current_instruction, MacroAssembler::is_mtctr(*current_instruction));
+
+    get_patchable_instruction_handle()->verify();
+    current_instruction += 2;
+
+    verify_op_code(current_instruction, Assembler::LWZ_OPCODE);
+
+    // cmpw (mnemonic)
+    verify_op_code(current_instruction, Assembler::CMP_OPCODE);
+
+    // bnectrl (mnemonic) (weak check; not checking the exact type)
+    verify_op_code(current_instruction, Assembler::BCCTR_OPCODE);
+
+    verify_op_code(current_instruction, Assembler::ISYNC_OPCODE);
+  }
+
+private:
+  static void verify_op_code_manually(uint*& current_instruction, bool result) {
+    assert(result, "illegal instruction sequence for nmethod entry barrier");
+    current_instruction++;
+  }
+  static void verify_op_code(uint*& current_instruction, uint expected,
+                             unsigned int mask = 63u << Assembler::OPCODE_SHIFT) {
+    // Masking both, current instruction and opcode, as some opcodes in Assembler contain additional information
+    // to uniquely identify simplified mnemonics.
+    // As long as the caller doesn't provide a custom mask, that additional information is discarded.
+    verify_op_code_manually(current_instruction, (*current_instruction & mask) == (expected & mask));
+  }
+};
+
+static NativeNMethodBarrier* get_nmethod_barrier(nmethod* nm) {
+  address barrier_address = nm->code_begin() + nm->frame_complete_offset() + (-9 * 4);
+
+  auto barrier = reinterpret_cast<NativeNMethodBarrier*>(barrier_address);
+  debug_only(barrier->verify());
+  return barrier;
+}
+
 void BarrierSetNMethod::deoptimize(nmethod* nm, address* return_address_ptr) {
-  ShouldNotReachHere();
+  // Nothing to do.
+  // Unlike other platforms, the frame resolution is done in the nmethod entry barrier stub.
+  // This way, writing frame information on the stack can be avoided.
```

Comment:
[optional] I'd replace the last sentence:
// PPC64 always has a valid back chain so the stub can simply pop the frame and there's nothing to do here.